src/java.desktop/share/classes/javax/swing/plaf/basic/BasicFileChooserUI.java

```diff
@@ -720,7 +721,8 @@ public void valueChanged(ListSelectionEvent evt) {
                                 boolean isDir = f.isDirectory();
                                 if ((chooser.isFileSelectionEnabled() && !isDir)
                                     || (chooser.isDirectorySelectionEnabled()
-                                        && (fsv.isFileSystem(f) || Files.isSymbolicLink(f.toPath()))
+                                        && (fsv.isFileSystem(f)
+                                        || (fsv.isLink(f) && Files.isSymbolicLink(f.toPath())))
```

Comment:
I understand that, virtual folders don't have a filesystem path, this is why `fsv.isFileSystem(f)` should handle this situation. The implementation of `FileSystemView.isFileSystem` explicitly excludes such files which are both `isLink` and `isDirectory`:
https://github.com/openjdk/jdk/blob/d7245f70e7bac1236bbcdcd9b25346ca22ab8bb2/src/java.desktop/share/classes/javax/swing/filechooser/FileSystemView.java#L382-L388

Let's look deeper:
https://github.com/openjdk/jdk/blob/d7245f70e7bac1236bbcdcd9b25346ca22ab8bb2/src/java.desktop/windows/classes/sun/awt/shell/Win32ShellFolder2.java#L591-L597

https://github.com/openjdk/jdk/blob/d7245f70e7bac1236bbcdcd9b25346ca22ab8bb2/src/java.desktop/windows/classes/sun/awt/shell/Win32ShellFolder2.java#L860-L866

The `hasAttribute` method uses `getAttributes0` which calls [`IShellFolder::GetAttributesOf`](https://learn.microsoft.com/en-us/windows/win32/api/shobjidl_core/nf-shobjidl_core-ishellfolder-getattributesof) which returns [`SFGAO`](https://learn.microsoft.com/en-us/windows/win32/shell/sfgao). The attribute `SFGAO_LINK` which corresponds to `ATTRIB_LINK` in Java is described as _“The specified items are shortcuts.”_ That is it is a shell folder shortcut with `.lnk` file extension, which used to be the case.

However, now this attribute is set for both [_symbolic links_](https://learn.microsoft.com/en-us/windows/win32/fileio/creating-symbolic-links) and [_junctions_](https://learn.microsoft.com/en-us/windows/win32/fileio/hard-links-and-junctions#junctions).

Here's what different methods return:
```text
file: C:\test\symbolic-links\jdk - junction
	isFileSystem   false
	isLink         true
	isDirectory    true
	isSymbolicLink false
Absolute Path : null

file: C:\test\symbolic-links\jdk - symbolic
	isFileSystem   false
	isLink         true
	isDirectory    true
	isSymbolicLink true
Absolute Path : C:\test\symbolic-links\jdk - symbolic

file: C:\test\symbolic-links\jdk - shortcut.lnk
	isFileSystem   false
	isLink         true
	isDirectory    true
	isSymbolicLink false
Absolute Path : null
```

In the three cases above, `isFileSystem` returns `false` even though all the three objects point to a filesystem directory, the first two are fully transparent for apps whereas the `.lnk` shortcut needs special handling.

If `JFileChooser` is in directory-only mode, selecting a junction or `.lnk` file results in selected file being to `null`; symbolic links are supported with your fix.

So, **I propose modifying the implementation of `FileSystemView.isFileSystem`** so that it returns only `sf.isFileSystem()` or filters out the `.lnk` shortcuts only.

Reply:
> I understand that, virtual folders don't have a filesystem path, this is why `fsv.isFileSystem(f)` should handle this situation. The implementation of `FileSystemView.isFileSystem` explicitly excludes such files which are both `isLink` and `isDirectory`:
> 
> https://github.com/openjdk/jdk/blob/d7245f70e7bac1236bbcdcd9b25346ca22ab8bb2/src/java.desktop/share/classes/javax/swing/filechooser/FileSystemView.java#L382-L388
> 
> Let's look deeper:
> 
> https://github.com/openjdk/jdk/blob/d7245f70e7bac1236bbcdcd9b25346ca22ab8bb2/src/java.desktop/windows/classes/sun/awt/shell/Win32ShellFolder2.java#L591-L597
> 
> https://github.com/openjdk/jdk/blob/d7245f70e7bac1236bbcdcd9b25346ca22ab8bb2/src/java.desktop/windows/classes/sun/awt/shell/Win32ShellFolder2.java#L860-L866
> 
> The `hasAttribute` method uses `getAttributes0` which calls [`IShellFolder::GetAttributesOf`](https://learn.microsoft.com/en-us/windows/win32/api/shobjidl_core/nf-shobjidl_core-ishellfolder-getattributesof) which returns [`SFGAO`](https://learn.microsoft.com/en-us/windows/win32/shell/sfgao). The attribute `SFGAO_LINK` which corresponds to `ATTRIB_LINK` in Java is described as _“The specified items are shortcuts.”_ That is it is a shell folder shortcut with `.lnk` file extension, which used to be the case.
> 
> However, now this attribute is set for both [_symbolic links_](https://learn.microsoft.com/en-us/windows/win32/fileio/creating-symbolic-links) and [_junctions_](https://learn.microsoft.com/en-us/windows/win32/fileio/hard-links-and-junctions#junctions).
> 
> Here's what different methods return:
> 
> ```
> file: C:\test\symbolic-links\jdk - junction
> 	isFileSystem   false
> 	isLink         true
> 	isDirectory    true
> 	isSymbolicLink false
> Absolute Path : null
> 
> file: C:\test\symbolic-links\jdk - symbolic
> 	isFileSystem   false
> 	isLink         true
> 	isDirectory    true
> 	isSymbolicLink true
> Absolute Path : C:\test\symbolic-links\jdk - symbolic
> 
> file: C:\test\symbolic-links\jdk - shortcut.lnk
> 	isFileSystem   false
> 	isLink         true
> 	isDirectory    true
> 	isSymbolicLink false
> Absolute Path : null
> ```
> 
> In the three cases above, `isFileSystem` returns `false` even though all the three objects point to a filesystem directory, the first two are fully transparent for apps whereas the `.lnk` shortcut needs special handling.
> 
> If `JFileChooser` is in directory-only mode, selecting a junction or `.lnk` file results in selected file being to `null`; symbolic links are supported with your fix.
> 
> So, **I propose modifying the implementation of `FileSystemView.isFileSystem`** so that it returns only `sf.isFileSystem()` or filters out the `.lnk` shortcuts only.

Since `FileSystemView.isFileSystem` is used in many places for particular Look and Feel, there might be chances of regression. So how about checking for shell folder first and if yes then can combine the `shell.isFileSystem()` with `isSymbolicLink()` in BasicFileChooser class.......? If its not a shell folder then directly we can check for `isSymbolicLink()`.

Reply:
As for other Look-and-Feels, `isFileSystem` is used for the same purpose that it's used in the code you're modifying: if a selected object in the list is a file system object (a directory), the text field to the path of the selected object; otherwise, set the text field to `null`.

Reply:
> Another option is to modify `isFileSystem` so that it returns `true` for all filesystem objects, excluding the added condition that it should be not a link to directory.

I assume this is required to ensure `JFileChooser` doesn't return `.lnk` file when in directory-only selection mode, and probably doesn't report is directory is selected when you select `.lnk` file. Double-clicking the `.lnk` file resolves to its target, then the directory can be safely returned.

Reply:
> So now two modifications are required.
> 
> 1. Modify `isLink` of win32ShellFolder class to return true only if its  `.lnk`.

This should be enough to fix this bug.

And this should enable the support for Windows junctions too.

Reply:
Yeah, this should do the work.

Reply:
To re-iterate:

> Currently, `isFileSystem` returns `false` when both `isLink` and `isDirectory` return `true`, which is the case for symbolic links and junctions.

This is the root cause why symbolic links need special handling.

So, if you make `isLink` return `false` for symbolic links and junctions, `isFileSystem` will return `true`. That's what we want: neither symbolic links nor junctions need special handling.